### PR TITLE
Cleanup & De-duplication for custom offset classes

### PR DESCRIPTION
**Pull request by jbrockmendel**
The diff looks a little bit mangled.  This PR does two things.

1) Centralize frequently-repeated calendar `__init__` in a new _CustomMixin

2) Merge a bunch of shared CustomBusinessMonthEnd and CustomBusinessMonthBegin logic into a parent class.

- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tseries/offsets.py [hunk 1]
```diff
@@ -413,6 +413,25 @@ def _from_name(cls, suffix=None):
         return cls()
 
 
+class _CustomMixin(object):
+    """
+    mixin for classes that define and validate calendar, holidays,
```
Comment (gfyoung) [inline pandas/tseries/offsets.py]: Capitalize "mixin"

Comment (gfyoung) [reply pandas/tseries/offsets.py]: There's one below `BusinessMixin` that you can also fix.

File: pandas/tseries/offsets.py [hunk 2]
```diff
@@ -572,9 +591,26 @@ def __init__(self, start='09:00', end='17:00', offset=timedelta(0)):
         kwds = {'offset': offset}
         self.start = kwds['start'] = _validate_business_time(start)
         self.end = kwds['end'] = _validate_business_time(end)
-        self.kwds = kwds
+        self.kwds.update(kwds)
         self._offset = offset
 
+    @cache_readonly
+    def next_bday(self):
+        # used for moving to next businessday
```
Comment (gfyoung) [inline pandas/tseries/offsets.py]: If you're going to discuss what the function does, let's use a doc-string instead of a comment.

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine. pls update docs as per @gfyoung comments.

File: pandas/tseries/offsets.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tseries/offsets.py]: This comment is retained from the existing version of the method, not sure it merits a full-on docstring.  But will do.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel 

pls document anything you can (you are already doing this), but more is better than less. yes doc-strings are boring :<, but this helps improve things even more.

**Comment by jbrockmendel** [comment]
> pls document anything you can (you are already doing this), but more is better than less. yes doc-strings are boring :<, but this helps improve things even more.

Will do.  Current gameplan vis-a-vis offsets:
  - fix confusing (several GH issues) [naming scheme](https://github.com/pandas-dev/pandas/issues/18235#issuecomment-346220122) in Quarter offsets, pending getting the "go-ahead"
  - hope/ask [for help](https://github.com/pandas-dev/pandas/pull/18224#issuecomment-348055959) with the cython-immutability goal, since that will entail a _major_ perf win.
  - update/fix docs on "anchored", implement a few missing non-anchored (Year, Quarter, Month, Day) offsets
  - distinguish between offsets that are Period-friendly vs just-for-date_range (there are a bunch of GH issues that boil down to confusion on this point)
- fix issues with Tick comparison methods (1+ GH issues)

**Pull request merged**